src/core/distributions/phylogenetics/tree/birthdeath/OccurrenceBirthDeathProcess.cpp

```diff
@@ -0,0 +1,543 @@
+
+    homogeneous_rho = dynamic_cast<const TypedDagNode<double >*>(ineventsampling);
+    addParameter( homogeneous_rho );
+
+
+    //check that lengths of vector arguments are sane
+    if ( heterogeneous_lambda != NULL && !(interval_times->getValue().size() == heterogeneous_lambda->getValue().size() - 1) )
+    {
+      throw(RbException("If provided as a vector, argument lambda must have one more element than timeline."));
+    }
+
+    if ( heterogeneous_mu != NULL && !(interval_times->getValue().size() == heterogeneous_mu->getValue().size() - 1) )
+    {
+      throw(RbException("If provided as a vector, argument mu must have one more element than timeline."));
+    }
+
+    if ( heterogeneous_psi != NULL && !(interval_times->getValue().size() == heterogeneous_psi->getValue().size() - 1) )
+    {
+      throw(RbException("If provided as a vector, argument psi must have one more element than timeline."));
+    }
+
+    if ( heterogeneous_r != NULL && !(interval_times->getValue().size() == heterogeneous_r->getValue().size() - 1) )
+    {
+      throw(RbException("If provided as a vector, argument r must have one more element than timeline."));
+    }
+
+    if ( heterogeneous_o != NULL && !(interval_times->getValue().size() == heterogeneous_o->getValue().size() - 1) )
+    {
+      throw(RbException("If provided as a vector, argument o must have one more element than timeline."));
+    }
+
+    updateVectorParameters();
+
+    if(timeline.size() > 1) {
+    RbVector<Clade> constr;
+    StartingTreeSimulator simulator;
+    RevBayesCore::Tree *my_tree = simulator.simulateTree( taxa, constr );
+
+    // store the new value
+    delete value;
+    value = my_tree;
+    }
+    else {
+      redrawValue();
+    }
+}
+
+
+
+
+
+/**
+ * The clone function is a convenience function to create proper copies of inherited objected.
+ * E.g. a.clone() will create a clone of the correct type even if 'a' is of derived type 'B'.
+ *
+ * \return A new copy of myself
+ */
+OccurrenceBirthDeathProcess* OccurrenceBirthDeathProcess::clone( void ) const
+{
+    return new OccurrenceBirthDeathProcess( *this );
+}
+
+
+
+
+
+/**
+ * Compute the log-transformed probability of the current value under the current parameter values.
+ */
+ double OccurrenceBirthDeathProcess::computeLnProbabilityDivergenceTimes( void )
+ {
+   std::cout << "computeLnProbabilityDivergenceTimes" << std::endl;
+
+     // variable declarations and initialization
+     double lnProbTimes = computeLnProbabilityTimes();
+
+     return lnProbTimes;
+ }
+
+
+
+
+
+/**
+ * Compute the log probability of the current value under the current parameter values.
+ */
+double OccurrenceBirthDeathProcess::computeLnProbabilityTimes( void ) const
+{
+  updateVectorParameters();
+
+  // compute the log-likelihood : use ComputeLikelihoodsBackwardsLt (backward traversal of the tree) or ComputeLikelihoodsForwardsMt (forward traversal of the tree)
+  const RevBayesCore::Tree tree(*value);
```

Comment:
This creates a deep copy and is slow. Perhaps better
const Tree& tree = *value;